src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,265 @@
+   * @param {!./layout-rect.LayoutRectDef} viewport.
+   * @param {./layout-rect.LayoutRectDef=} opt_container. relative to viewport
+   */
+  tick(viewport, opt_container) {
+    // Normalize container LayoutRect to be relative to page
+    if (opt_container) {
+      opt_container =
+          moveLayoutRect(opt_container, viewport.left, viewport.top);
+    }
+    let elementRect;
+    let ownerRect = null;
+
+    if (this.isAmpElement_) {
+      // If calls with container, always assume getLayoutBox() return relative
+      // LayoutRect to container
+      elementRect = this.element_.getLayoutBox();
+      const owner = this.element_.getOwner();
+      ownerRect = owner && owner.getLayoutBox();
+      if (opt_container) {
+        elementRect = moveLayoutRect(elementRect, opt_container.left,
+            opt_container.top);
+        if (ownerRect) {
+          ownerRect = moveLayoutRect(ownerRect, opt_container.left,
+              opt_container.top);
+        }
+      }
+    } else {
+      // If the element is not an AMP element, always assume its position to
+      // the page won't change.
+      if (!this.elementRectForNonAMP_) {
+        elementRect = this.element_./*OK*/getBoundingClientRect();
+        const normRect = opt_container || viewport;
+        this.elementRectForNonAMP_ =
+            moveLayoutRect(elementRect, normRect.left, normRect.top);
+      }
+      elementRect = this.elementRectForNonAMP_;
+    }
+
+    // Normalize container element
+    const changeEntry = this.getValidIntersectionChangeEntry_(
+        elementRect, ownerRect, viewport, opt_container);
+    if (changeEntry) {
+      this.callback_(changeEntry);
+    }
+  }
+
+  /**
+   * Return a change entry for that should be compatible with
+   * IntersectionObserverEntry if it's valid with current config.
+   *
+   * Mutates passed in rootBounds to have x and y according to spec.
+   *
+   * @param {!./layout-rect.LayoutRectDef} element element's rect
+   * @param {?./layout-rect.LayoutRectDef} owner element's owner rect
+   * @param {!./layout-rect.LayoutRectDef} viewport viewport's rect.
+   * @param {./layout-rect.LayoutRectDef=} opt_container.
+   * @return {?IntersectionObserverEntry} A valid change entry.
+   * @private
+   */
+  getValidIntersectionChangeEntry_(element, owner, viewport, opt_container) {
+    // TODO: need to include owner LayoutRect info
+
+    // Building an IntersectionObserverEntry.
+    let intersectionRect = element;
+    const nonIntersectRect = layoutRectLtwh(0, 0, 0, 0);
+    if (owner) {
+      intersectionRect = rectIntersection(owner, element) || nonIntersectRect;
+    }
+    if (opt_container) {
+      intersectionRect = rectIntersection(opt_container, intersectionRect) ||
+          nonIntersectRect;
+    }
+    intersectionRect = rectIntersection(viewport, intersectionRect) ||
+        nonIntersectRect;
+
+    intersectionRect = moveLayoutRect(intersectionRect, -viewport.left,
```

Comment:
Can still be [moved later](https://github.com/ampproject/amphtml/pull/6100#discussion_r87512537).